### DISCUSSION TITLE
Refactoring things

### DISCUSSION THREAD

**Discussion by nicewook** [nicewook/corona-visual-server#6] [General]
@kkweon 님께 질문

당장 개선을 이야기하고픈게 아니고 PR주신 리팩터링을 보며 느꼈던 바를 나누고, 의견을 들어보고 싶어서 글을 남깁니다. 
소소한 것들도 있습니다. 

1. SeoulTZ
제가 익숙하지 않은지 처음에 뭔지 헷갈렸습니다. SeoulTimeZone 이라고 조금 늘여쓰는게 가독성이 좋아보입니다. 
이런 작은 것들도 일일이 별도의 PR을 하는것이 맞을까요? PR 잘 활용하는 현업에서는 어떤가 궁금합니다.

2. config
config를 정리하고 사용하는 측(= main)에서 설정을 모아서 보고, 그것을 필요로 하는 곳으로 보내주는 의미로 이해했습니다. 좋다고 생각합니다. 
다만, 연관이 없는 필드들이 하나의 config 안에 모여있는게 어색하긴 했습니다. 예를 들어 fetcher에는 필요없는 config들이 함께 전달이 되는 것이 불편하게 느껴졌네요. 작은 프로젝트이니 문제없고, 확장되고 필요성이 느껴지면 변경하면 되려나요? 

3. config on runtime
이 경우는 아니지만 런타임에서 변경되는 config 정보들이 있다면 어떻게 처리하는게 좋을까요? 우선 떠오르는건 config 패키지 내에 값들을 저장하고 RWMutex로 get, set을 해주면 어떨까 싶네요

4. handler에 fetcher 넣기 
지금은 handler가 하나 뿐이고, 그게 fetcher를 필요로 하지만 handler가 늘어나는 경우에는 어떻게 구현하는게 좋을까요? 
예를 들어 handler가 10개가 되고, 각각이 자신만의 fetcher 또는 다른 기능함수가 필요하다면? 

미리 감사드립니다. :-)
가볍게 떠오르는 대로 생각을 나눠주셔도 도움 되겠습니다.

**Comment by kkweon** [comment]
이런 디스커션 좋습니다 😄 

아래는 저의 생각이니 현석님의 생각과는 얼마든지 다를 수 있습니다.

1. 현업에서는 필요하다고 생각합니다. 말씀하신 것처럼 모두에게 익숙하고 일관성이 중요한 문제인데 어 다른데서는 이렇게 쓰는데? 라고 미쳐 발견 못했을 수 도 있으니까요.
2. XP 방법론대로 따를 것 같습니다. iterative하게 계속 발전시켜 나가는 것이죠. 제가 보기엔 config 자체가 서비스키 빼고는 전부 필요 없어 보입니다. 리팩토링을 쉽게 하기 위해서 일단 묶어 놓게 되었네요. 예를 들어 날짜포맷(YYYYMMDD)은 데이터에 종속된 것이죠. 날짜 포맷이 클라이언트 요청에 맞게 바뀌는것이 아니라요. XML 을 전부 string이 아니라 제대로 타입 프로그래밍을 했으면 예를 들어 `customTimeFormat` 을 만들어서 처리했으면 좋을 것 같습니다. 이거는 PR 로 보여드리겠습니다. 그동안 수정하실 게 있으시면 마음대로 하셔도 되구요.
3. 이거는 요구 사항에 따라 맞추어 갈 것 같습니다. 말씀하신것도 될 수도 있고 단순히 싱글톤으로 될 수 도 있구요. 아니면 그냥 레디스나 제대로 된 데이터베이스에게 맡길 수 도 있구요 (예를 들어 재실행되도 config가 유지되어야 하나요? 등등). 실제 프로덕션에서는 아예 config 전용 서비스가 있는 경우도 흔하니까요.
4. 이때도 마찬가지로 저는 계속 발전시켜 나가는 방법을 사용합니다. 말씀하신 것들이 전부 마이크로한 디자인 초이스들이어서 테스트코드만 작성한다면 뭐가 되든 될 것 같습니다. 보통 리포지토리 패턴과 서비스구조를 만들겠지만 코드로 봐야 되겠죠. 잘만든 API 라이브러리를 사용해보셨다면 아시겠지만 아무 생각없이 매우 쉽게 쓸 수 있습니다. 그렇게 결국 되도록 하는 것이 목표입니다.

대충 아시겠지만 저는 XP를 어느 정도 좋아하고 있어서 미래에 필요 없는 기능보다는 현재에 집중하고 테스트코드를 통해 iterative 하게 발전가능성을 열어두는 것을 중요시 합니다. 물론 큰 그림도 중요하지만요.

**Reply by nicewook** [reply]
감사히 읽었습니다. 머리속에 어렴풋하던게 개운해지네요. 
XP답게 현실에 충실해야겠습니다.
